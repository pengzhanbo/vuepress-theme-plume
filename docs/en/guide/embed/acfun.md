---
title: AcFun Video
icon: lets-icons:video-fill
createTime: 2025/10/08 10:57:45
permalink: /en/guide/embed/video/acfun/
badge: New
---

## Overview

The theme provides the capability to embed AcFun videos.

This feature is powered by [vuepress-plugin-md-power](../../config/plugins/markdown-power.md).

## Configuration

This feature is disabled by default. You need to enable it in the theme configuration.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      acfun: true, // [!code ++]
    },
  })
})
```

## Syntax

Simple syntax:

```md
@[acfun](id)
```

More options:

```md
@[acfun width="100%" height="400px" ratio="16:9"](id)
```

**Parameter Description:**

- id: Video ID
- width: Video width
- height: Video height
- ratio: Video aspect ratio, defaults to `16:9`

## Examples

### Widescreen Video

Input:

```md
@[acfun](ac47431669)
```

Output:

@[acfun](ac47431669)
