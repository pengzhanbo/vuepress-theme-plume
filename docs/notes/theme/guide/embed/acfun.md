---
title: AcFun 视频
icon: lets-icons:video-fill
createTime: 2025/06/28 10:57:45
permalink: /guide/embed/video/acfun/
badge: 新
---

## 概述

主题提供了 嵌入 AcFun 视频 的功能。

该功能由 [vuepress-plugin-md-power](../../config/plugins/markdown-power.md) 提供支持。

## 配置

该功能默认不启用。你需要在主题配置中开启。

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      acfun: true, // [!code ++]
    },
  })
})
```

## 语法

简单的语法：

```md
@[acfun](id)
```

更多选项：

```md
@[acfun width="100%" height="400px" ratio="16:9"](id)
```

**选项说明：**

- id: 视频 ID
- width: 视频宽度
- height: 视频高度
- ratio: 视频比例，默认 `16:9`

## 示例

### 宽频视频

输入：

```md
@[acfun](ac47431669)
```

输出：

@[acfun](ac47431669)
