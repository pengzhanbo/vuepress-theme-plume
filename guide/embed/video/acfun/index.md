---
url: /guide/embed/video/acfun/index.md
---
## 概述

主题提供了 嵌入 AcFun 视频 的功能。

该功能由 [vuepress-plugin-md-power](../../config/plugins/markdown-power.md) 提供支持。

## 配置

该功能默认不启用。你需要在主题配置中开启。

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      acfun: true, // [!code ++]
    },
  })
})
```

## 语法

简单的语法：

```md
@[acfun](id)
```

更多选项：

```md
@[acfun width="100%" height="400px" ratio="16:9"](id)
```

**选项说明：**

* id: 视频 ID
* width: 视频宽度
* height: 视频高度
* ratio: 视频比例，默认 `16:9`

## 示例

### 宽频视频

输入：

```md
@[acfun](ac47431669)
```

输出：

@[acfun](ac47431669)
