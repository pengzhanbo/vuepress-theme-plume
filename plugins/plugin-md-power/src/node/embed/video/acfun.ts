/**
 * @[acfun](acid)
 */
import type { PluginWithOptions } from 'markdown-it'
import type { AcFunTokenMeta } from '../../../shared/index.js'
import { parseRect } from '../../utils/parseRect.js'
import { resolveAttrs } from '../../utils/resolveAttrs.js'
import { stringifyAttrs } from '../../utils/stringifyAttrs.js'
import { createEmbedRuleBlock } from '../createEmbedRuleBlock.js'

const AC_FUN_LINK = 'https://www.acfun.cn/player'

export const acfunPlugin: PluginWithOptions<never> = (md) => {
  createEmbedRuleBlock<AcFunTokenMeta>(md, {
    type: 'acfun',
    name: 'video_acfun',
    syntaxPattern: /^@\[acfun([^\]]*)\]\(([^)]*)\)/,
    meta([, info, id]) {
      const { attrs } = resolveAttrs(info)

      return {
        id,
        title: attrs.title || 'AcFun',
        width: attrs.width ? parseRect(attrs.width) : '100%',
        height: attrs.height ? parseRect(attrs.height) : undefined,
        ratio: attrs.ratio ?? '16:10',
      }
    },
    content(meta) {
      const { id, width, height, ratio, title } = meta
      const src = `${AC_FUN_LINK}/${id}`
      return `<VPVideoEmbed${stringifyAttrs({ src, width, height, ratio, title, type: 'acfun' })} />`
    },
  })
}
