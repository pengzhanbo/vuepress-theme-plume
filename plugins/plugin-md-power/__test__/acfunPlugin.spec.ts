import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { acfunPlugin } from '../src/node/embed/video/acfun.js'

function createMarkdown() {
  return MarkdownIt().use((md) => {
    md.block.ruler.before('code', 'import_code', () => false)
    md.renderer.rules.import_code = () => ''
  }).use(acfunPlugin)
}

describe('acfunPlugin', () => {
  it('should work', () => {
    const md = createMarkdown()
    const code = `\
@[acfun](123456)

@[acfun title="test"](123456)

@[acfun width="100%" height="600px"](123456)

@[acfun width="100%" ratio="16:9"](123456)
`

    expect(md.render(code)).toMatchSnapshot()
  })

  it('should not work', () => {
    const md = createMarkdown()
    const code = `\
@[acfun]()

@[acfun]xxx

@[ acfun]123456

@[ acfun](123456)
`

    expect(md.render(code)).toMatchSnapshot()
  })
})
