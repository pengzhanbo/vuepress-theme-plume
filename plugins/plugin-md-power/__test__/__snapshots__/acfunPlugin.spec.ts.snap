// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`acfunPlugin > should not work 1`] = `
"<VPVideoEmbed src="https://www.acfun.cn/player/" width="100%" ratio="16:10" title="AcFun" type="acfun" /><p>@[acfun]xxx</p>
<p>@[ acfun]123456</p>
<p>@<a href="123456"> acfun</a></p>
"
`;

exports[`acfunPlugin > should work 1`] = `"<VPVideoEmbed src="https://www.acfun.cn/player/123456" width="100%" ratio="16:10" title="AcFun" type="acfun" /><VPVideoEmbed src="https://www.acfun.cn/player/123456" width="100%" ratio="16:10" title="test" type="acfun" /><VPVideoEmbed src="https://www.acfun.cn/player/123456" width="100%" height="600px" ratio="16:10" title="AcFun" type="acfun" /><VPVideoEmbed src="https://www.acfun.cn/player/123456" width="100%" ratio="16:9" title="AcFun" type="acfun" />"`;
